openedx/core/lib/block_structure/transformer.py

```diff
@@ -125,4 +126,55 @@ def transform(self, usage_info, block_structure):
                 block structure, with already collected data for the
                 transformer, that is to be transformed in place.
         """
-        pass
+        raise NotImplementedError
+
+
+class OptimizedTransformer(BlockStructureTransformer):
+    """
+    Transformers may optionally choose to subclass this class instead of the
+    base, if their transform logic can be broken apart into a lambda for
+    optimization of combined tree traversals.
+
+    For performance reasons, developers should try to subclass this
+    class instead of the above base class, whenever possible - since
+    with this alternative, traversal of the entire block structure happens
+    only once for all transformers that implement this form of transform.
+    """
+
+    def transform(self, usage_info, block_structure):
```

Comment:
Ok.  Leave as is then.  I can see the value in allowing callers to always assume that the `transform` method is available for all Transformers when/if needed.

Currently, though, all clients always go through the higher layer BlockStructure classes, like Transformers and the Manager.  So what transform method is used is an implementation detail.  As you can see, we didn't need to update any unit tests for a single Transformer.